### PR TITLE
Implement FastPAM1. Fixes #106

### LINKED ISSUE #106: Implement FastPAM1, and cite PAM and FastPAM papers
Title. See #1 for context.

We need to implement the FastPAM1 optimization to reduce the complexity of the ground truth algorithm to `O(n^2)`. 

Additionally, we should include references to the PAM and FastPAM papers in the comments of those functions.

### PR DESCRIPTION
**Pull request by mailology**
I have implemented `FastPAM1` algorithm in this commit. The `FastPAM1` algorithm follows from Algorithm 3 of the PAM and FastPAM1 paper https://arxiv.org/pdf/1810.05691.pdf, where the core idea is to replace the iteration of `k` medoids in the `swap` step with an if conditional outside the loop. This is achievable by using an array to compute `k` loss changes which results in O(N^2) per iteration. 

The algorithm is tested against MNIST-1k dataset as shown below.

Screen Shot 2021-08-05 at 12 47 17 AM

It gives the expected set of medoids and the order of `k` is close of zero.

### CODE REVIEW

File: src/fastpam1.cpp [hunk 1]
```diff
@@ -0,0 +1,193 @@
+/**
```
Comment (motiwari) [inline src/fastpam1.cpp:1]: Include header guards, like [here](https://github.com/ThrunGroup/BanditPAM/blob/b2bfaea933f32ad4ca6acc922e68baf6daa59500/headers/kmedoids_algorithm.hpp#L1-L2)

File: headers/kmedoids_algorithm.hpp [hunk 2]
```diff
@@ -185,10 +185,16 @@ class KMedoids {
 
     void fit_naive(arma::mat inputData);
 
+    void fit_fastpam1(arma::mat inputData);
```
Comment (motiwari) [inline headers/kmedoids_algorithm.hpp:188]: Move below `swap_naive`

File: setup.py [hunk 3]
```diff
@@ -274,7 +274,8 @@ def build_extensions(self):
         [os.path.join('src', 'kmedoids_pywrapper.cpp'), 
          os.path.join('src', 'kmedoids_algorithm.cpp'),
          os.path.join('src', 'pam.cpp'),
-         os.path.join('src', 'banditpam.cpp')],
+         os.path.join('src', 'banditpam.cpp'),
+         os.path.join('src', 'fastpam1.cpp')],
```
Comment (motiwari) [inline setup.py:278]: Add a trailing comma after `)`, then put `],` on a newline (to preserve blame)

File: src/fastpam1.cpp [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+ * @file fastpam1.cpp
```
Comment (motiwari) [inline src/fastpam1.cpp:2]: As per [the issue](https://github.com/ThrunGroup/BanditPAM/issues/106), include citations to Schubert's original PAM paper and the FastPAM paper

File: src/fastpam1.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+#include <armadillo>
+#include <unordered_map>
+#include <regex>
+
+/**
+ * \brief Runs naive FastPAM1 algorithm.
+ *
+ * Run the naive FastPAM1 algorithm to identify a dataset's medoids.
+ *
+ * @param input_data Input data to find the medoids of
```
Comment (motiwari) [inline src/fastpam1.cpp:20]: `s/to find the medoids of/to cluster`

File: src/fastpam1.cpp [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+ */
+void KMedoids::fit_fastpam1(arma::mat input_data) {
+  data = input_data;
+  data = arma::trans(data);
+  arma::rowvec medoid_indices(n_medoids);
+  // runs build step
```
Comment (motiwari) [inline src/fastpam1.cpp:26]: rm

File: src/fastpam1.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  KMedoids::build_fastpam1(data, medoid_indices);
+  steps = 0;
```
Comment (motiwari) [inline src/fastpam1.cpp:28]: appears unused; rm?

File: src/fastpam1.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+  while (iter < max_iter && medoidChange) {
+    auto previous(medoid_indices);
+    // runs swap step as necessary
+    KMedoids::swap_fastpam1(data, medoid_indices, assignments);
+    medoidChange = arma::any(medoid_indices != previous);
+    iter++;
+  }
+  medoid_indices_final = medoid_indices;
+  labels = assignments;
+  steps = iter;
```
Comment (motiwari) [inline src/fastpam1.cpp:42]: appears unused; rm?

File: src/fastpam1.cpp [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+  // runs build step
+  KMedoids::build_fastpam1(data, medoid_indices);
+  steps = 0;
+  medoid_indices_build = medoid_indices;
+  arma::rowvec assignments(data.n_cols);
+  size_t iter = 0;
+  bool medoidChange = true;
+  while (iter < max_iter && medoidChange) {
+    auto previous(medoid_indices);
+    // runs swap step as necessary
```
Comment (motiwari) [inline src/fastpam1.cpp:35]: rm

File: src/fastpam1.cpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+  arma::rowvec medoid_indices(n_medoids);
+  // runs build step
+  KMedoids::build_fastpam1(data, medoid_indices);
+  steps = 0;
+  medoid_indices_build = medoid_indices;
+  arma::rowvec assignments(data.n_cols);
+  size_t iter = 0;
+  bool medoidChange = true;
+  while (iter < max_iter && medoidChange) {
+    auto previous(medoid_indices);
```
Comment (motiwari) [inline src/fastpam1.cpp:34]: I think `{brace initialization}` is usually preferred; correct me if I'm wrong

File: src/fastpam1.cpp [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    // runs swap step as necessary
+    KMedoids::swap_fastpam1(data, medoid_indices, assignments);
+    medoidChange = arma::any(medoid_indices != previous);
+    iter++;
+  }
+  medoid_indices_final = medoid_indices;
+  labels = assignments;
```
Comment (motiwari) [inline src/fastpam1.cpp:41]: appears unused; rm?

File: src/fastpam1.cpp [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+/**
+ * \brief Build step for the FastPAM1 algorithm
+ *
+ * Runs build step for the FastPAM1 algorithm. Loops over all datapoint and
+ * checks its distance from every other datapoint in the dataset, then checks if
+ * the total cost is less than that of the medoid (if a medoid exists yet).
+ *
+ * @param data Transposed input data to find the medoids of
```
Comment (motiwari) [inline src/fastpam1.cpp:52]: `s/to find the medoids of/to cluster`

File: src/fastpam1.cpp [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+  bool use_absolute = true;
+  arma::rowvec estimates(N, arma::fill::zeros);
+  arma::rowvec best_distances(N);
+  best_distances.fill(std::numeric_limits<double>::infinity());
+  arma::rowvec sigma(N); // standard deviation of induced losses on reference points
+  for (size_t k = 0; k < n_medoids; k++) {
+    double minDistance = std::numeric_limits<double>::infinity();
+    int best = 0;
+    KMedoids::build_sigma(
+           data, best_distances, sigma, batchSize, use_absolute); // computes std dev amongst batch of reference points
```
Comment (motiwari) [inline src/fastpam1.cpp:71]: rm comment

File: src/fastpam1.cpp [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+          cost = best_distances(j);
+        }
+        total += cost;
+      }
+      if (total < minDistance) {
+        minDistance = total;
+        best = i;
+      }
+    }
+    // update the medoid index for that of lowest cost
```
Comment (motiwari) [inline src/fastpam1.cpp:89]: rm

File: src/fastpam1.cpp [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+  }
+}
+
+/**
+ * \brief Swap step for the FastPAM1 algorithm
+ *
+ * Runs swap step for the FastPAM1 algorithm. Loops over all datapoint and
+ * checks its distance from every other datapoint in the dataset, then checks if
+ * the total cost is less than that of the medoid.
+ *
```
Comment (motiwari) [inline src/fastpam1.cpp:113]: Update comment

File: src/fastpam1.cpp [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+ * @param data Transposed input data to find the medoids of
```
Comment (motiwari) [inline src/fastpam1.cpp:114]: `s/to find the medoids of/to cluster` throughout

File: src/fastpam1.cpp [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+ * that is modified in place as better medoids are identified
+ * @param assignments Uninitialized array of indices corresponding to each
+ * datapoint assigned the index of the medoid it is closest to
+ */
+void KMedoids::swap_fastpam1(
+  arma::mat& data, 
+  arma::rowvec& medoid_indices,
+  arma::rowvec& assignments)
+{
+  double bestChange = 0; // best loss change
```
Comment (motiwari) [inline src/fastpam1.cpp:125]: rm comment

File: src/fastpam1.cpp [hunk 18]
```diff
@@ -0,0 +1,193 @@
...
+  // write the sigma distribution to logfile
+  sigma_log(sigma);
+  // for every point in our dataset, let it serve as a new medoid 
+  for (size_t i = 0; i < data.n_cols; i++) {
+      double di = best_distances(i);
+      // loss change for making i a medoid
+      delta_td.fill(-di);
+      for (size_t j = 0; j < data.n_cols; j++) {
+          if (j != i) {
+              // compute distance to new medoid
```
Comment (motiwari) [inline src/fastpam1.cpp:156]: rm

**Review by motiwari — CHANGES_REQUESTED** [review]
Nice work! Definitely a nontrivial change. 

Nits throughout.

File: src/fastpam1.cpp (hunk 8 above)
Comment (mailology) [reply src/fastpam1.cpp:46]: We use it to count the number of swap steps and it is used here:
https://github.com/ThrunGroup/BanditPAM/blob/67221ff929d8005e46358a9151795b66bce66891/src/kmedoids_algorithm.cpp#L294-L301

File: src/fastpam1.cpp (hunk 11 above)
Comment (mailology) [reply src/fastpam1.cpp:45]: `labels` is a global variable which stores the assignment of medoids for each datapoint.
https://github.com/ThrunGroup/BanditPAM/blob/67221ff929d8005e46358a9151795b66bce66891/src/kmedoids_algorithm.cpp#L89-L97

**Comment by mailology** [comment]
Fixed the nits for [PR#112 ](https://github.com/ThrunGroup/BanditPAM/pull/112). The code is retested and it works as expected.

File: src/fastpam1.cpp (hunk 7 above)
Comment (motiwari) [reply src/fastpam1.cpp:33]: Ah; that function uses a lot of globals. Instead, we should make those member variables. It's ok for this PR and I've updated in #110 for us to fix later.

File: src/fastpam1.cpp (hunk 11 above)
Comment (motiwari) [reply src/fastpam1.cpp:45]: OK for this PR; updated #110 to fix

File: src/fastpam1.cpp (hunk 8 above)
Comment (motiwari) [reply src/fastpam1.cpp:46]: OK for this PR; updated #110 to fix

**Review by motiwari — CHANGES_REQUESTED** [review]
Suggestions:
- Create `fastpam1.hpp`
- Cite original PAM paper

Everything else OK

**Comment by mailology** [comment]
The followings are the changes I have made:

- Cite original PAM paper.
- Introduce `fastpam1.hpp` and create `FastPAM1` subclass inherited from `KMedoids`. There are three member functions in this child: `fit_fastpam1`, `build_fastpam1` and `swap_fastpam1` (they are removed in `KMedoids` class). The definition of these three functions is in `fastpam1.cpp`.
- The usual way we call fit function: `this->*fitFn` seems unnecessarily complicated. In fact, the pointer `this` is from a parent object in the perspective of the fit function `fit_fastpam1` and we need to modify the pointer in order to point to a child function. Also, it turns out that `fitFn` is an intermediate variable which is not needed at all. I have modified the logic so that whenever we do `fit`, the pointer directly points to the corresponding fit function (similar update will be made on `naive` and `BanditPAM` in another PR). 
- I have modified `km::KMedoids::checkAlgorithm` so that it only does checking of the algorithm since we don't need to deal with `fitFn` anymore.

The updated code is checked against Python implementation of `naive`, `BanditPAM` and `FastPAM1` algorithms. All give expected results on MNIST-1k data.

**Review by motiwari — APPROVED** [review]
Looks great, nice work @mailology !

**Pull request merged**